pennylane/qnode.py

```diff
@@ -410,6 +373,163 @@ def construct(self, args, kwargs=None):
         #: dict[int->str]: map from free parameter index to the gradient method to be used with that parameter
         self.grad_method_for_par = {k: self._best_method(k) for k in self.variable_ops}
 
+    def construct_metric_tensor(self, args, **kwargs):
+        """Create metric tensor subcircuits for each parameter.
+
+        If the parameter appears in a gate :math:`G`, the subcircuit contains
+        all gates which precede :math:`G`, and :math:`G` is replaced by the variance
+        value of its generator.
+
+        Args:
+            args (tuple): Represent the free parameters passed to the circuit.
+                Here we are not concerned with their values, but with their structure.
+                Each free param is replaced with a :class:`~.variable.Variable` instance.
+
+        Keyword Args:
+            diag_approx (bool): If ``True``, forces the diagonal
+                approximation. Default is ``False``.
+
+        .. note::
+
+            Additional keyword arguments may be passed to the quantum circuit function, however PennyLane
+            does not support differentiating with respect to keyword arguments. Instead,
+            keyword arguments are useful for providing data or 'placeholders' to the quantum circuit function.
+        """
+        # pylint: disable=too-many-statements
+        diag_approx = kwargs.pop("diag_approx", False)
+
+        if not self.ops or not self.cache:
+            # construct the circuit
+            self.construct(args, kwargs)
+
+        # convert the queue to a DAG
+        G = to_DiGraph(self.queue, self.ev)
+
+        # keep track of the layer number
+        layer = 0
+        layer_ops = {0: {"ops": [], "pidx": []}}
+
+        variable_ops_sorted = sorted(list(self.variable_ops.items()), key=lambda x: x[1][0][0])
+
+        for param_idx, gate_param_tuple in variable_ops_sorted:
```

Comment:
This whole DAG building / building of the layers will go into the `QNode` core soon. It is good to have some working code already present for this :+1:

But I would consider to create our own object-oriented implementation of a DAG specifically taylored to our purpose instead of using networkx: 
* We use only a very small subset of networkx's functionality (I think you included it mostly for the topological sorting?)
* We will have much more readable code because we don't have to carry around abstract graph objects and access them through the networkx API
* We should also be able to do without the topological sorting because this is mostly an artifact of using networkx, isn't it? The user will already enter the gates in a topologically sorted order (an order he might actually care about)
* Layering can be integrated in the implementation and not be added on top of it

What are your thoughts about this?

Reply:
In cases like this, where there already exists a very highly maintained and well-used package for a particular purpose (here, graph representation and algorithms), imo it's best to make use of them rather than try and re-invent the wheel. In particular,

* NetworkX is a pure-Python library, so doesn't cause any binary dependency issues

* It avoids us having to maintain and test a DAG graph builder that could have bugs/edge cases that have already been solved by the NetworkX team

* The different graph algorithms used within NetworkX are likely much more optimized that ones we would write on first iteration. For example, under the hood, the `networkx.dag.ancestors` function used below already makes use of a variety of shortest path length algorithms depending on the structure of the input graph.

Reply:
I agree though that it would be nice to have layering integrated in the implementation and not be added on top of it. Perhaps we should split the auto-layering done here as its own utility function?

Reply:
Okay, what about making a `Circuit` (or other name) class that encapsulates the DAG, the queue and the layers? I would like to have all `networkx` code staying in that class so that it is properly wrapped.

This would give us an API like this
```python
circuit = Circuit(queue) # could also be built en route

circuit.layers[1] -> Array(Command) # all commands in a layer in queued order
circuit.queue -> Array(Command) # all commands in the queue in user order
circuit.wire -> Array(Command) # all commands on a single wire
circuit.graph -> nx.DAG(Command) # all commands as DAG
```
The class `Command` would then wrap around `Operation` and have methods like `direct_predecessors`, `all_predecessors` etc. that handle the traversing of the DAG.

Reply:
I really like this idea! This will help massively with the DAG abstraction.